src/modules/systemlib/param/param.c

```diff
@@ -1139,14 +1143,18 @@ param_export(int fd, bool only_unsaved)
 out:
 
 	if (result == 0) {
-		result = bson_encoder_fini(&encoder);
-
-		// write and finish
-		if ((result != 0) || write(fd, encoder.buf, encoder.bufpos) != encoder.bufpos) {
-			PX4_ERR("param write error");
+		if (write(fd, encoder.buf, encoder.bufpos) == encoder.bufpos) {
+			// switch encoder back to fd and finalize
```

Comment:
I don't like this. This is a hack to work around tinybson-specific internals. I think we should move all of this logic (including buffering) inside tinybson, and add support for `bson_encoder_init_buffered_file(bson_encoder_t encoder, int fd, void *buf, unsigned bufsize)`. That should not be too hard, and `bson_encoder_s` already contains all the necessary fields.

Reply:
I agree, let's do it properly.